File: typescript/src/n0226_invert_binary_tree.ts
import type { TreeNode } from './utils/tree';

export function invertTree(root: TreeNode | null): TreeNode | null {
  if (!root) {
    return null;
  }
  const stack = [root];
  while (stack.length > 0) {
    const node = stack.pop()!;
    [node.left, node.right] = [node.right, node.left];
    if (node.left) {
      stack.push(node.left);
    }
    if (node.right) {
      stack.push(node.right);
    }
  }
  return root;
}

export function invertTreeRecursive(root: TreeNode | null): TreeNode | null {
  if (!root) {
    return null;
  }
  [root.left, root.right] = [
    invertTreeRecursive(root.right),
    invertTreeRecursive(root.left),
  ];
  return root;
}

File: python/src/n0226_invert_binary_tree.py
from typing import Optional
from utils.tree import TreeNode


def invert_tree(root: Optional[TreeNode]) -> Optional[TreeNode]:
    if not root:
        return None
    stack = [root]
    while stack:
        node = stack.pop()
        node.left, node.right = node.right, node.left
        if node.left:
            stack.append(node.left)
        if node.right:
            stack.append(node.right)
    return root


def invert_tree_recursive(root: Optional[TreeNode]) -> Optional[TreeNode]:
    if not root:
        return None
    root.left, root.right = (
        invert_tree_recursive(root.right),
        invert_tree_recursive(root.left),
    )
    return root
